### ISSUE TITLE
String option without value and required param fail

### ISSUE THREAD

**Issue by razvanh** [remkop/picocli#979]
I have the current setup for oprions and parameters:

```
@CommandLine.Parameters(arity = "1", description = "The file.")
	private File file;

@CommandLine.Option(names = {"-w"}, arity="0..1", description = "W Option")
	private String delimiters;
```

If I run my program with `nameOfUtility.java path/to/file` or `nameOfUtility.java -w "" path/to/file` or `nameOfUtility.java -w "whatever" path/to/file` everything works.

This however does not, it throws a `MissingParameterException`:

`nameOfUtility.java -w path/to/file`

I tried using `defaultValue` and `fallbackValue` for the option but the result is always the same. Am I wrong in expecting the above to work ?

**Comment by remkop** [comment]
TL;DR: yes, this is how picocli currently handles options with an optional parameter. 

The picocli parser is “greedy” when it handles optional parameters: it looks at the value after `-w`, and if it’s likely to be a parameter then it will process the value as a parameter for `-w`.

The logic to determine if a value is “likely to be a parameter” checks if the value is a known subcommand and whether the value is (likely) another option, based on the prefix. For example, `-w -v path/to/file`, would be processed ok if your app also has a `-v` option. 

In your example, the `path/to/file` value could be a parameter for `-w`, so it is treated as one. There is currently no mechanism for backtracking and trying if an alternative interpretation would be able to process the input without resulting in an error. 

I can look into the feasibility of implementing such a backtracking mechanism. (May take a while though.)

**Comment by razvanh** [comment]
Sounds good! Appreciate the fast response. I wasn't sure if it was something I was doing wrong.

**Comment by remkop** [comment]
Glad to hear that!
I created https://github.com/remkop/picocli/issues/980 as a follow-up item.

**Comment by remkop** [comment]
Can I ask: how are you finding picocli so far? I'm always looking for things to improve, feedback very welcome!

**Comment by kyeo138** [comment]
Hi! Just wanted to say that I'm new to java, and after exploring apache commons, I love using picocli! It's so much easier to understand and use, especially for beginners. 

A question, as I may be facing the same issue as razvanh...
The main problem is that for me, the last parameter in the option string will always be the filename - no matter what's in front of it.  

Do you see a way to code around this issue?

**Comment by razvanh** [comment]
@kyeo138 Hey Jake - you don't have to parse all arguments with `picocli`. If you know that the last argument is always the file, you can only parse the other arguments with `picocli` and deal with that one manually. Hope this helps.

**Comment by razvanh** [comment]
@remkop Hey Remko - I think the library is pretty straightforward and easy to use, I don't have any specific feedback. Thanks!

**Comment by remkop** [comment]
> Hey Remko - I think the library is pretty straightforward and easy to use, I don't have any specific feedback. Thanks!

@razvanh No worries. Glad you like picocli. GitHub stars are always welcome. 😉

**Comment by remkop** [comment]
@kyeo138 Glad you like picocli.

> The main problem is that for me, the last parameter in the option string will always be the filename - no matter what's in front of it.

I'm guessing that you have an option with arity `1..*`, or `0..*`, in addition to a positional parameter. Something like this, perhaps?

```java
class Cmd {
  @Option(names = "-x", arity = "1..*")
  List<String> optionValues;
  
  @Parameter
  File file;
}
```
(Please let me know your actual command if I'm incorrect.)

If this is your command definition, then yes, depending on how users specify the values, the `file` may or may not not get a value.   If users specify `-x a b c path/to/file`, then the file will be added to the list with option values for `-x`. This is not what you want.

Side note: sometimes, we may get lucky, and the file gets a value if the end user specifies values a certain way, for example `path/to/file -x a b c` (file specified before the option) or `-x a b c -- path/to/file` (sophisticated user uses the "--" [end-of-options](https://picocli.info/#_double_dash) delimiter).
 

# Solution

I would suggest removing `arity = "1..*"`, and using the  [default arity](https://picocli.info/#_default_arity) intead. For a multi-value option (like a `List` or array) this is `1` (one). That means that users can specify the option multiple times to add a single value to the list. For example:

```java
class Unambiguous {
  @Option(names = "-x")
  List<String> optionValues;
  
  @Parameter
  File file;
}
```

Now users can specify input like this:

```
-x a -xb -x=c -x d path/to/file
```

This will result in the values `[a, b, c, d]` in the `optionValues` list, and the positional parameter correctly getting the `path/to/file` value.

What you can also do is allow users to specify multiple comma-separated values in a single option parameter. To accomplish this, use `split = ","` in the option definition. For example:

```java
class Unambiguous {
  @Option(names = "-x", split = ",")
  List<String> optionValues;
  
  @Parameter
  File file;
}
```

Now users can specify input like this:

```
-x a,b,c -x=d,e,f path/to/file
```

This will result in the values `[a, b, c, d, e, f]` in the `optionValues` list, and the positional parameter getting the `path/to/file` value.

Did this answer your question?

**Comment by kyeo138** [comment]
Yes thank you! I was able to strip the filename off the arguments and process the rest just fine.

**Comment by remkop** [comment]
Glad to hear that!

Remember, your GitHub stars for picocli are the only thing that fills the yawning void in my soul! :stuck_out_tongue_winking_eye:

**Comment by remkop** [comment]
I added section [Options with an Optional Parameter](https://picocli.info/#_options_with_an_optional_parameter) to the user manual to clarify the current behaviour.
